### ISSUE TITLE
Using for Python code and coverage.py

### ISSUE THREAD

**Issue by aodj** [coverallsapp/github-action#4]
I was just looking into this for use in the GitHub actions beta, and covering my Python code, but it looks like the coverage tooling for Python doesn't support `lcov` output (ie running `coverage.py` via `pytest-cov` on `pytest` unittests).

I know that the `coveralls` Python library can interface with the coveralls.io API, so I'm curious what the suggested solution might be.

Any ideas?

**Comment by nickmerwin** [comment]
Hi @aodj, could you try the steps described here:

https://github.com/coveralls-clients/coveralls-python/blob/master/docs/usage/multilang.rst

e.g.:

```bash
coveralls --output=coverage.json
coveralls-lcov -v -n coverage.info > coverage.json
```

Then use "./coverage.json" as the `path-to-lcov` input for the Coveralls Action.

Let me know how it goes, thanks!

**Comment by aodj** [comment]
I'll give it a try, but is installing and running a Ruby gem the only option here? I've previously used a workflow on Travis that simply invokes the `coveralls` tool which uploads the output from `coverage.py`, which is nice since it's simply a case of installing an extra Python lib.

**Comment by aodj** [comment]
I think I'm missing something; the `coveralls-lcov` tool doesn't output LCOV formatted data. It consumes LCOV data to output `json` [Looking at the `coveralls` Python library](https://github.com/coveralls-clients/coveralls-python/blob/master/coveralls/api.py#L167-L180), is it not possible to use the `api/v1/jobs` endpoint since that appears to consume the more common JSON format instead of LCOV:

**Comment by aodj** [comment]
Any more thoughts on using the action with Python?

**Comment by aodj** [comment]
Just to keep this ticket up-to-date I gave up with this action, and just used the [`coveralls`](https://pypi.org/project/coveralls/) Python package. [Workflow for reference](https://github.com/aodj/icelandreview/blob/master/.github/workflows/pythonpackage.yml#L30).

**Comment by mikkokotila** [comment]
@aodj dead simple approach solves the issue. Which raises an important question; how come the official version from the company itself, does not work?

**Comment by hwagit** [comment]
However, running just `coveralls` on a `pull_request` event does not prompt Coveralls.io to later on to add a comment to the PR, even though it is enabled in the Coveralls.io settings?

**Comment by abhinavsingh** [comment]
This probably should be re-opened or at-least put instructions for Python integration upfront in README.

**Comment by jvesely** [comment]
Running just `coveralls` in github actions reports branch `HEAD`. I ended up with:
```
    - name: Coveralls
      run: |
        CI_BRANCH=${GITHUB_REF#"ref/heads"} coveralls
      shell: bash
```
@hakonanes, There's also `CI_PULL_REQUEST` env var that you can set to something from `${{ github.event }}` context or parse it out of the file at `$GITHUB_EVENT_PATH`

**Comment by ly-cultureiq** [comment]
if https://github.com/nickmerwin/node-coveralls/pull/243 gets merged - i have a simple pr to add pytest and coverage.py support

**Comment by bboe** [comment]
For anyone following this issue, I was able to get parallel builds to work with `python-coveralls` and using this action just for the `parallel-finish` step.

See my workflow modification here: https://github.com/praw-dev/prawtools/commit/bb0e00b89a899fb2494cf230c0104746ddaeb098

Note getting this to work does depend on the changes here: https://github.com/coveralls-clients/coveralls-python/pull/216/files

**Comment by agronholm** [comment]
Just to un-confused others reading this: @bboe did not actually use [python-coveralls](https://pypi.org/project/python-coveralls/) (which is a separate client for coveralls.io), but a modified version of coveralls-python (available as [coveralls](https://pypi.org/project/coveralls/) on the PyPI).

This combination worked for me too. Let's hope the PR is merged.

**Comment by agronholm** [comment]
@bboe has your PR actually been merged? I tried with coveralls 2.0.0 and, unlike your version, it didn't work without using the coveralls token instead of the github token due to the 422 Unprocessable Entity error which is mentioned in the docs. And then, trying to use the official action to finish the parallel build, it fails due to an error like `No build matching CI build number 111541185 found`.

**Comment by agronholm** [comment]
I'm getting the same error even with the exact same version I used to use before, so I'm guessing something has changed upstream. Now I have absolutely no clue how to get this to work.

**Comment by MarvinT** [comment]
@agronholm For anyone still trying to get pytest working, I think there have been some changes to the env vars needed to run coveralls and get it to find the correct repo on coveralls.io
```
      - name: Coveralls
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          coveralls
```
N.B. I have to ```pip install git+https://github.com/TimoRoth/coveralls-python.git``` until https://github.com/coveralls-clients/coveralls-python/pull/227 is pulled in.

An example configuration is here: https://github.com/greenelab/ponyo/pull/11
but its complicated a bit by using a conda installation to deal with R dependencies

**Comment by casperdcl** [comment]
looks like https://github.com/AndreMiras/coveralls-python-action is a viable replacement for https://github.com/coverallsapp/github-action/ and https://github.com/AndreMiras/coveralls-python-action/pull/5 will even allow parallel builds.

**Comment by willronchetti** [comment]
Did something change recently? [This](https://github.com/4dn-dcic/Submit4DN/blob/master/.github/workflows/main.yml#L44) workflow was working with coveralls as of 12/14/2020 ([build](https://github.com/4dn-dcic/Submit4DN/runs/1552575847?check_suite_focus=true#step:5:204)), yet [now](https://github.com/4dn-dcic/Submit4DN/runs/1710591466?check_suite_focus=true#step:5:204) it does not?

```
Submitting coverage to coveralls.io...
Could not submit coverage: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs
resubmitting with id None-4337388734212009414
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/coveralls/api.py", line 252, in wear
    response.raise_for_status()
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/coveralls/cli.py", line 85, in main
    result = coverallz.wear()
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/coveralls/api.py", line 255, in wear
    raise CoverallsException('Could not submit coverage: {}'.format(e))
coveralls.exception.CoverallsException: Could not submit coverage: 422 Client Error: Unprocessable Entity for url: https://coveralls.io/api/v1/jobs
```

**Comment by scottgigante** [comment]
This works for me:
```
    - name: Coveralls
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        COVERALLS_SERVICE_NAME: github
      run: |
        coveralls
```

**Comment by afinetooth** [comment]
Hi @scottgigante.

Just FYI for anyone else reading: 

You are correct about those environment variables. Due to the nature of GitHub App authentication, the Coveralls GitHub Action is the only Coveralls integration that must use the repo's GitHub token, rather than its Coveralls-assigned repo token, to identify the repo to the Coveralls API. The way to "trick" GitHub Actions into doing the same for other integrations is to pass `COVERALLS_SERVICE_NAME=github`, which is the reserved value for the Coveralls GitHub Action.

In other words, the Coveralls API thinks you're using the Coveralls GitHub Action and uses the GitHub token to ID your repo.

Note that `v2.X` will soon support python and other coverage formats due to using a different underlying integration, the Coveralls Universal Coverage Reporter.
